src/shims/fs.rs

```diff
@@ -257,6 +261,181 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                e
+            ))?;
+        // `dirfd` should be a `c_int` but the `syscall` function provides an `isize`.
+        let dirfd: i32 = this
+            .read_scalar(dirfd_op)?
+            .to_machine_isize(&*this.tcx)?
+            .try_into()
+            .map_err(|e| err_unsup_format!(
+                "Failed to convert pointer sized operand to integer: {}",
+                e
+            ))?;
+        // we only support interpreting `path` as an absolute directory or as a directory relative
+        // to `dirfd` when the latter is `AT_FDCWD`. The behavior of `statx` with a relative path
+        // and a directory file descriptor other than `AT_FDCWD` is specified but it cannot be
+        // tested from `libstd`. If you found this error, please open an issue reporting it.
+        if !(path.is_absolute() || dirfd == this.eval_libc_i32("AT_FDCWD")?)
+        {
+            throw_unsup_format!(
+                "Using statx with a file descriptor different from `AT_FDCWD` is not supported"
+            )
+        }
+
+        // the `_mask_op` paramter specifies the file information that the caller requested.
+        // However `statx` is allowed to return information that was not requested or to not
+        // return information that was requested. This `mask` represents the information we can
+        // actually provide in any host platform.
+        let mut mask =
+            this.eval_libc("STATX_TYPE")?.to_u32()? | this.eval_libc("STATX_SIZE")?.to_u32()?;
+
+        // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
+        // symbolic links.
+        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
+            // FIXME: metadata for symlinks need testing.
+            std::fs::symlink_metadata(path)
+        } else {
+            std::fs::metadata(path)
+        };
+
+        let metadata = match metadata {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                return Ok(-1);
+            }
+        };
+
+        let file_type = metadata.file_type();
+
+        // The `mode` field specifies the type of the file and the permissions over the file for
+        // the owner, its group and other users. Given that we can only provide the file type
+        // without using platform specific methods, we only set the bits corresponding to the file
+        // type. This should be an `__u16` but `libc` provides its values as `u32`.
+        let mode = if file_type.is_file() {
+            this.eval_libc("S_IFREG")?
+        } else if file_type.is_dir() {
+            this.eval_libc("S_IFDIR")?
+        } else {
+            this.eval_libc("S_IFLNK")?
+        }.to_u32()?;
+
+        let mode: u16 = mode.try_into().map_err(|e|
+            err_unsup_format!("Failed to convert 32-bit value to u16: {}",
```

Comment:
That's not a very informative error. What is that 32bit value? And what is the type of these `S_IFREG` etc constants anyway?

Reply:
That's the thing. They are `mode_t` which is `u32` according to `libc` but the type of `mode` in `man statx` is `__u16`.

Reply:
Well it'd ICE if they are not `i32`... so seems like we got some `i32` constants that are actually all `u16` values.

So I think you can `unwrap` here, we don't support bad libc... or `.expect("libc contains bad value for S_* constant")` or so to keep some context.